book-content/chapters/15-designing-your-types.md

```diff
@@ -0,0 +1,1444 @@
+
+## Conditional Types
+
+You can use conditional types in TypeScript to create if/else logic in your types. This is mostly useful in a library setting when working with really complex code, but I'll show you a simple example in case you ever run into it.
+
+Let's imagine we create a `ToArray` generic type that converts a type to an array type:
+
+```tsx
+type ToArray<T> = T[];
+```
+
+This is fine, except when we pass in a type that's already an array. If we do, we'll get an array of arrays:
+
+```ts twoslash
+type ToArray<T> = T[];
+// ---cut---
+type Example = ToArray<string>;
+//   ^?
+
+type Example2 = ToArray<string[]>;
+//   ^?
+```
+
+We actually want `Example2` to end up as `string[]` too. So, we'll need to check if `T` is already an array, and if it is, we'll return `T` instead of `T[]`.
+
+We can use a conditional type to do that. This uses a ternary operator, similar to JavaScript:
+
+```tsx
+type ToArray<T> = T extends any[] ? T : T[];
+```
+
+This will look pretty scary the first time you see it, but let's break it down.
+
+```tsx
+type ToArray<T> = T extends any[] ? T : T[];
+//                ^^^^^^^^^^^^^^^   ^   ^^^
+//                condition       true/false
+```
+
+### The Condition
+
+The 'condition' in a conditional type is the part before the `?`. In this case, it's `T extends any[]`.
+
+```tsx
+type ToArray<T> = T extends any[] ? T : T[];
+//                ^^^^^^^^^^^^^^^
+//                   condition
+```
+
+This checks if `T` can be assigned to `any[]`. To make sense of this check, imagine it like a function:
+
+```tsx
+const toArray = (t: any[]) => {
+  // implementation
+};
+```
+
+What could be passed to this function? Only arrays:
+
+```ts twoslash
+// @errors: 2345
+const toArray = (t: any[]) => {
+  // implementation
+};
+
+// ---cut---
+toArray([1, 2, 3]); // OK
+toArray("hello");
+```
+
+`T extends any[]` checks if `T` could be passed to a function expecting `any[]`. If we wanted to check if `T` was a string, we'd use `T extends string`.
+
+### 'True' and 'False'
+
+```tsx
+type ToArray<T> = T extends any[] ? T : T[];
+//                                  ^   ^^^
+//                                 true/false
+```
+
+If the condition is true, it resolves to the 'true' part, just like a normal ternary. If it's false, it resolves to the 'false' part.
+
+In this case, if `T` is an array, it resolves to `T`. If it's not, it resolves to `T[]`.
+
+This means that our examples above now work as expected:
+
+```ts twoslash
+type ToArray<T> = T extends any[] ? T : T[];
+
+// ---cut---
+type Example = ToArray<string>;
+//   ^?
+
+type Example2 = ToArray<string[]>;
+//   ^?
+```
+
+Conditional types turn TypeScript's type system into a full programming language. They're incredibly powerful, but they can also be incredibly complex. You'll rarely need them in application code, but they can perform wonders in library code.
+
+## Mapped Types
```

Comment:
it could be worthwhile to mention that homomorphic mapped types can also map over arrays/tuples and preserve their arrayness - it would be a nice comparison to show what happens with arrays/tuples when u pass them through non-homomorphic mapped types